ast.c

```diff
@@ -724,6 +720,50 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
 			ast->kind = ZEND_AST_ASSIGN_OP;
 			ast->attr = AST_BINARY_COALESCE;
 			break;
+		case ZEND_AST_CLASS_NAME:
+			if (state->version < 70) {
+				zval name_zval;
+				ast_to_zval(&name_zval, ast->child[0], state);
+				zval class_name_zval;
+				ast_create_virtual_node_ex(
+						&class_name_zval, AST_NAME, ast->child[0]->attr, zend_ast_get_lineno(ast), state, 1, &name_zval);
+				zval const_zval;
+				ZVAL_STR(&const_zval, AST_STR(str_class));
+				ast_create_virtual_node_ex(
+						zv, ZEND_AST_CLASS_CONST, 0, zend_ast_get_lineno(ast), state, 2, &class_name_zval, &const_zval);
+				return;
+			}
+			break;
+#else
+		case ZEND_AST_CLASS_CONST:
+			if (state->version >= 70) {
+				zend_ast *class_name_ast = ast->child[0];
+				if (class_name_ast->kind != ZEND_AST_ZVAL) {
+					break;
+				}
+				// Convert to an AST_CLASS_NAME instead. The opposite of the work done in the ZEND_AST_CLASS_NAME case.
+				zend_ast *const_name_ast = ast->child[1];
+				if (const_name_ast->kind != ZEND_AST_ZVAL) {
+					break; // Currently impossible
```

Comment:
If it's impossible, we should drop the check. `zend_ast_get_zval` contains an assertion that checks this. Even better would be to use `zend_ast_get_str()` to directly get at the zend_string.

Reply:
```
php > var_export([]::class);

Fatal error: Cannot use ::class with dynamic class name in php shell code on line 1
```

Okay, it's possible (I just didn't test every possible edge case) - that's a runtime error, not a syntax error

Reply:
Okay, I'm going to have to report a bug with php 7.4-dev debug ZTS on bugs.php.net:

```php
php > var_export((2)::class);                                                                                                                                                                  
php: /path/to/php-src/Zend/zend_ast.h:298: zend_ast_get_str: Assertion `zval_get_type(&(*(zv))) == 6' failed.
[1]    4949 abort      php -a
```

Reply:
Filed https://bugs.php.net/bug.php?id=77530&thanks=4

Reply:
It turns out that I was mistaken and it indeed something I should check for (I tried `2::class` the first time but not `(2)::class`)